### ISSUE TITLE
Error running atuin daemon

### ISSUE THREAD

**Issue by raffaelewylde** [atuinsh/atuin#2289]
### What did you expect to happen?

expected atuin daemon to startup

### What happened?

I know the daemon isn't completely implemented yet. Figured I'd let you guys know about this issue. I've tried running the daemon every so often fo a couple weeks and always get the same error.
```
 atuin daemon                     
Error: No such file or directory (os error 2)

Location:
    /home/wylde/.cargo/registry/src/index.crates.io-6f17d22bba15001f/atuin-daemon-0.2.0/src/server.rs:206:10
```
and yet that file exists...
```
 ls  /home/wylde/.cargo/registry/src/index.crates.io-6f17d22bba15001f/atuin-daemon-0.2.0/src/ 
server/  client.rs  history.rs  lib.rs  server.rs*
```
The contents of server.rs are 

```
use eyre::WrapErr;

use atuin_client::encryption;
use atuin_client::history::store::HistoryStore;
use atuin_client::record::sqlite_store::SqliteStore;
use atuin_client::settings::Settings;
use std::path::PathBuf;
use std::sync::Arc;
use time::OffsetDateTime;
use tracing::{instrument, Level};

use atuin_client::database::{Database, Sqlite as HistoryDatabase};
use atuin_client::history::{History, HistoryId};
use dashmap::DashMap;
use eyre::Result;
use tonic::{transport::Server, Request, Response, Status};

use crate::history::history_server::{History as HistorySvc, HistoryServer};

use crate::history::{EndHistoryReply, EndHistoryRequest, StartHistoryReply, StartHistoryRequest};

mod sync;

#[derive(Debug)]
pub struct HistoryService {
    // A store for WIP history
    // This is history that has not yet been completed, aka a command that's current running.
    running: Arc<DashMap<HistoryId, History>>,
    store: HistoryStore,
    history_db: HistoryDatabase,
}

impl HistoryService {
    pub fn new(store: HistoryStore, history_db: HistoryDatabase) -> Self {
        Self {
            running: Arc::new(DashMap::new()),
            store,
            history_db,
        }
    }
}

#[tonic::async_trait()]
impl HistorySvc for HistoryService {
    #[instrument(skip_all, level = Level::INFO)]
    async fn start_history(
        &self,
        request: Request<StartHistoryRequest>,
    ) -> Result<Response<StartHistoryReply>, Status> {
        let running = self.running.clone();
        let req = request.into_inner();

        let timestamp =
            OffsetDateTime::from_unix_timestamp_nanos(req.timestamp as i128).map_err(|_| {
                Status::invalid_argument(
                    "failed to parse timestamp as unix time (expected nanos since epoch)",
                )
            })?;

        let h: History = History::daemon()
            .timestamp(timestamp)
            .command(req.command)
            .cwd(req.cwd)
            .session(req.session)
            .hostname(req.hostname)
            .build()
            .into();

        // The old behaviour had us inserting half-finished history records into the database
        // The new behaviour no longer allows that.
        // History that's running is stored in-memory by the daemon, and only committed when
        // complete.
        // If anyone relied on the old behaviour, we could perhaps insert to the history db here
        // too. I'd rather keep it pure, unless that ends up being the case.
        let id = h.id.clone();
        tracing::info!(id = id.to_string(), "start history");
        running.insert(id.clone(), h);

        let reply = StartHistoryReply { id: id.to_string() };

        Ok(Response::new(reply))
    }

    #[instrument(skip_all, level = Level::INFO)]
    async fn end_history(
        &self,
        request: Request<EndHistoryRequest>,
    ) -> Result<Response<EndHistoryReply>, Status> {
        let running = self.running.clone();
        let req = request.into_inner();

        let id = HistoryId(req.id);

        if let Some((_, mut history)) = running.remove(&id) {
            history.exit = req.exit;
            history.duration = match req.duration {
                0 => i64::try_from(
                    (OffsetDateTime::now_utc() - history.timestamp).whole_nanoseconds(),
                )
                .expect("failed to convert calculated duration to i64"),
                value => i64::try_from(value).expect("failed to get i64 duration"),
            };

            // Perhaps allow the incremental build to handle this entirely.
            self.history_db
                .save(&history)
                .await
                .map_err(|e| Status::internal(format!("failed to write to db: {e:?}")))?;

            tracing::info!(
                id = id.0.to_string(),
                duration = history.duration,
                "end history"
            );

            let (id, idx) =
                self.store.push(history).await.map_err(|e| {
                    Status::internal(format!("failed to push record to store: {e:?}"))
                })?;

            let reply = EndHistoryReply {
                id: id.0.to_string(),
                idx,
            };

            return Ok(Response::new(reply));
        }

        Err(Status::not_found(format!(
            "could not find history with id: {id}"
        )))
    }
}

#[cfg(unix)]
async fn shutdown_signal(socket: Option<PathBuf>) {
    let mut term = tokio::signal::unix::signal(tokio::signal::unix::SignalKind::terminate())
        .expect("failed to register sigterm handler");
    let mut int = tokio::signal::unix::signal(tokio::signal::unix::SignalKind::interrupt())
        .expect("failed to register sigint handler");

    tokio::select! {
        _  = term.recv() => {},
        _  = int.recv() => {},
    }

    eprintln!("Removing socket...");
    if let Some(socket) = socket {
        std::fs::remove_file(socket).expect("failed to remove socket");
    }
    eprintln!("Shutting down...");
}

#[cfg(windows)]
async fn shutdown_signal() {
    tokio::signal::windows::ctrl_c()
        .expect("failed to register signal handler")
        .recv()
        .await;
    eprintln!("Shutting down...");
}

#[cfg(unix)]
async fn start_server(settings: Settings, history: HistoryService) -> Result<()> {
    use tokio::net::UnixListener;
    use tokio_stream::wrappers::UnixListenerStream;

    let socket_path = settings.daemon.socket_path;

    let (uds, cleanup) = if cfg!(target_os = "linux") && settings.daemon.systemd_socket {
        #[cfg(target_os = "linux")]
        {
            use eyre::OptionExt;
            tracing::info!("getting systemd socket");
            let listener = listenfd::ListenFd::from_env()
                .take_unix_listener(0)?
                .ok_or_eyre("missing systemd socket")?;
            listener.set_nonblocking(true)?;
            let actual_path = listener
                .local_addr()
                .context("getting systemd socket's path")
                .and_then(|addr| {
                    addr.as_pathname()
                        .ok_or_eyre("systemd socket missing path")
                        .map(|path| path.to_owned())
                });
            match actual_path {
                Ok(actual_path) => {
                    tracing::info!("listening on systemd socket: {actual_path:?}");
                    if actual_path != std::path::Path::new(&socket_path) {
                        tracing::warn!(
                            "systemd socket is not at configured client path: {socket_path:?}"
                        );
                    }
                }
                Err(err) => {
                    tracing::warn!("could not detect systemd socket path, ensure that it's at the configured path: {socket_path:?}, error: {err:?}");
                }
            }
            (UnixListener::from_std(listener)?, false)
        }
        #[cfg(not(target_os = "linux"))]
        unreachable!()
    } else {
        tracing::info!("listening on unix socket {socket_path:?}");
        (UnixListener::bind(socket_path.clone())?, true)
    };

    let uds_stream = UnixListenerStream::new(uds);
    Server::builder()
        .add_service(HistoryServer::new(history))
        .serve_with_incoming_shutdown(
            uds_stream,
            shutdown_signal(cleanup.then_some(socket_path.into())),
        )
        .await?;
    Ok(())
}

#[cfg(not(unix))]
async fn start_server(settings: Settings, history: HistoryService) -> Result<()> {
    use tokio::net::TcpListener;
    use tokio_stream::wrappers::TcpListenerStream;

    let port = settings.daemon.tcp_port;
    let url = format!("127.0.0.1:{}", port);
    let tcp = TcpListener::bind(url).await?;
    let tcp_stream = TcpListenerStream::new(tcp);

    tracing::info!("listening on tcp port {:?}", port);

    Server::builder()
        .add_service(HistoryServer::new(history))
        .serve_with_incoming_shutdown(tcp_stream, shutdown_signal())
        .await?;
    Ok(())
}

// break the above down when we end up with multiple services

/// Listen on a unix socket
/// Pass the path to the socket
pub async fn listen(
    settings: Settings,
    store: SqliteStore,
    history_db: HistoryDatabase,
) -> Result<()> {
    let encryption_key: [u8; 32] = encryption::load_key(&settings)
        .context("could not load encryption key")?
        .into();

    let host_id = Settings::host_id().expect("failed to get host_id");
    let history_store = HistoryStore::new(store.clone(), host_id, encryption_key);

    let history = HistoryService::new(history_store.clone(), history_db.clone());

    // start services
    tokio::spawn(sync::worker(
        settings.clone(),
        store,
        history_store,
        history_db,
    ));

    start_server(settings, history).await
}
```

### Atuin doctor output

```yaml
Checking for diagnostics


Please include the output below with any bug reports or issues

{
  "atuin": {
    "version": "18.3.0",
    "sync": {
      "cloud": true,
      "records": true,
      "auto_sync": true,
      "last_sync": "2024-07-17 6:10:20.060512026 +00:00:00"
    },
    "sqlite_version": "3.44.0"
  },
  "shell": {
    "name": "zsh",
    "default": "zsh",
    "plugins": [],
    "preexec": null
  },
  "system": {
    "os": "Parrot Security",
    "arch": "x86_64",
    "version": "6.1",
    "disks": [
      {
        "name": "/dev/sdb3",
        "filesystem": "btrfs"
      },
      {
        "name": "/dev/sdb3",
        "filesystem": "btrfs"
      },
      {
        "name": "/dev/sdb1",
        "filesystem": "vfat"
      }
    ]
  }
}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ellie** [comment]
The error is actually from that file, on line 206. Just Rust panics can be a bit weird to read sometimes. It's complaining because it cannot find the path to the daemon socket file

Does changing 

```
[daemon]
socket_path = "/path/that/exists/atuin.sock"
```

work for you?

**Comment by pyinto** [comment]
Had a similar issue a spent a bunch of time solving it, here some info that might help you:

- relative path doesn't work, so neither this 
`"~/.local/share/atuin/atuin.sock"`
or that
`"$HOME/.local/share/atuin/atuin.sock"`
work for me. Only full path.
- try to remove `atuin.sock` file, I did it but not sure it did anything.

**Comment by fmuehlis** [comment]
I'm also having this issue on every system I use Atuin on (NixOS + home-manager and Ubuntu 20.04 with home-manager). It only happens on startup for me where I have to solve it by:

1. Run `atuin daemon` again in the terminal, which fails with:
```shell
Error: failed to connect to local atuin daemon. Is it running?

Location:
    /build/source/crates/atuin-daemon/src/client.rs:31:26
Error: Address already in use (os error 98)

Location:
    /build/source/crates/atuin-daemon/src/server:206:10
``` 
3. Delete the socket `rm ~/.local/share/atuin/atuin.sock`
4. Run `autin daemon` and now it works, if I then stop it manually the issue does not come back.

I'm using this via nix and home-manager with this config:
```nix
atuin = {
      enable = true;
      package = pkgs.unstable.atuin;
      enableFishIntegration = true;
      enableBashIntegration = true;
      enableNushellIntegration = true;
      settings = {
        sync_address = "<my-url>";
        key_path = config.sops.secrets."<atuin-key-file>".path;
        daemon = {
          enabled = true;
        };
      };
    };
``` 

I had to switch to the daemon because I'm now using ZFS on the NixOS machine. The Ubuntu installation runs on btrfs though. If needed I can provide more information.

**Comment by WieeRd** [comment]
This seems to happen when Atuin daemon has been stopped abruptly before.
Normally, Atuin daemon deletes the socket file upon terminating.
If it fails to do so, the socket remains, and it will cause "Address already in use" error on the next execution of Atuin.

To reproduce:

1. Run `atuin daemon`
2. While the daemon is running, force-quit it with `pkill -9 "atuin"`
  (`ps`, `pgrep` or whatever to make sure you are killing the right thing)
3. `ls ~/.local/share/atuin/atuin.sock` will show that the socket file still remains
4. Running `atuin daemon` again will fail with `Error: Address already in use (os error 98)`
5. `rm ~/.local/share/atuin/atuin.sock`
6. `atuin daemon` works again

**Comment by WieeRd** [comment]
As a band-aid fix to prevent this problem, add `rm -f ~/.local/share/atuin/atuin.sock` before `atuin daemon` in whatever script or service that autostarts atuin daemon, to ensure that the socket is cleaned up before starting a new daemon.

In the future, Atuin should be made to properly handle the case when the socket file already exists.

**Comment by nir-megery** [comment]
Not sure if it matters in this case, but the standard way, according to [bind#notes](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/bind.2.html), would be to call [unlink](https://crosvm.dev/doc/base/sys/unix/struct.UnlinkUnixListener.html) before binding.

**Comment by Comnir** [comment]
Not sure if it matters in this case, but the standard way, according to [bind#notes](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/bind.2.html), would be to call [unlink](https://crosvm.dev/doc/base/sys/unix/struct.UnlinkUnixListener.html) before binding.

**Comment by dougg0k** [comment]
Here, if I set just. 

```
[daemon]
enabled = true
```

It does not work. Every command that I type, this comes as the prefix.

```
Error: failed to connect to local atuin daemon at /home/user/.local/share/atuin/atuin.sock. Is it running?

Caused by:
   0: transport error
   1: No such file or directory (os error 2)
   2: No such file or directory (os error 2)

Location:
    /home/runner/work/atuin/atuin/crates/atuin-daemon/src/client.rs:36:14
Error: nu::shell::non_zero_exit_code

  × External command had a non-zero exit code
    ╭─[/home/user/.local/share/atuin/init.nu:17:34]
 16 │     if not ($cmd | str starts-with $ATUIN_KEYBINDING_TOKEN) {
 17 │         $env.ATUIN_HISTORY_ID = (atuin history start -- $cmd)
    ·                                  ──┬──
    ·                                    ╰── exited with code 1
 18 │     }
    ╰────
```

In this case pointing to the `.local` path, only because I had also set `socket_path = "/home/user/.local/share/atuin/atuin.sock"` which only works as a full path.

If `socket_path` are not set. It defaults to ` /run/user/1000/atuin.sock.` with the same error above.

However if I manually run `atuin daemon` with `enabled = false` it works.

**Comment by neitsab** [comment]
@ellie suggested I enable the daemon in [this comment](https://github.com/atuinsh/atuin/issues/1726#issuecomment-2669867909) on another issue. That's supposed to be for ZFS but well, when the software's author suggests something you better try it 😊

So I edited my config to use

```toml
[daemon]
enabled = true
```

then ran `atuin daemon`. Here's the result:

```
Error: failed to connect to local atuin daemon at /run/user/1000/atuin.sock. Is it running?

Caused by:
   0: transport error
   1: No such file or directory (os error 2)
   2: No such file or directory (os error 2)

Location:
    /usr/src/debug/atuin/atuin-18.4.0/crates/atuin-daemon/src/client.rs:36:14
^CRemoving socket...
Shutting down...
```

I tried in a new terminal tab but no dice.
 
`atuin doctor` output:

```json
{
  "atuin": {
    "version": "18.4.0",
    "sync": {
      "cloud": true,
      "records": true,
      "auto_sync": true,
      "last_sync": "2025-02-19 22:13:22.345613869 +00:00:00"
    },
    "sqlite_version": "3.46.0"
  },
  "shell": {
    "name": "zsh",
    "default": "zsh",
    "plugins": [
      "atuin"
    ],
    "preexec": "built-in"
  },
  "system": {
    "os": "Arch Linux",
    "arch": "x86_64",
    "version": "unknown",
    "disks": [
      {
        "name": "/dev/mapper/root",
        "filesystem": "btrfs"
      },
      {
        "name": "/dev/mapper/root",
        "filesystem": "btrfs"
      },
      {
        "name": "/dev/nvme1n1p1",
        "filesystem": "vfat"
      }
    ]
  }
}
```

I just want to contribute some data towards the improvement of the "daemon workaround".

Cheers

**Comment by neilbags** [comment]
Still a bug when the socket path is a relative path (~/.local/share/atuin/atuin.sock) as is given in the example in the config file.

Setting an absolute path works

**Comment by ArpanSriv** [comment]
I ended up using netcat to listen on a socket and using that in the atuin config.toml:

```
nc -lkU /path/to/sock.sock
```

This seems to have worked, but I don't like this solution at all :)

**Comment by bwhaley** [comment]
For anyone coming across this on a fresh install in MacOS Tahoe, I edited `~/.config/atuin/config.toml` to set an absolute path for `data_dir`, e.g.

```
data_dir = "/Users/bwhaley/.local/share/atuin"
```
